File: src/game/entities/tower.cpp
#include "tower.h"

Tower::Tower()
{
    this->id = "";
    this->damagePerShoot = 10;
    this->attackSpeed = 1000;
    this->xp = 0;
    this->type = 1;
    this->level = 1;
    this->baseAttackSpeed = 1000;
    this->baseDamagePerShoot = 1;
    this->xpForLevelUp = 20;
    this->i = 0;
    this->j = 0;
}

Tower::Tower(QString id, int damagePerShoot, int attackSpeed, int xp, int type, int level, int baseAttackSpeed, int baseDamagePerShoot, int xpForLevelUp, int i, int j)
{
    this->id = id;
    this->damagePerShoot = damagePerShoot;
    this->attackSpeed = attackSpeed;
    this->xp = xp;
    this->type = type;
    this->level = level;
    this->baseAttackSpeed = baseAttackSpeed;
    this->baseDamagePerShoot = baseDamagePerShoot;
    this->xpForLevelUp = xpForLevelUp;
    this->i = i;
    this->j = j;
}

void Tower::write(QJsonObject &jsonObj) const
{
    jsonObj["id"] = id;
    jsonObj["damagePerShot"] = damagePerShoot;
    jsonObj["attackSpeed"] = attackSpeed;
    jsonObj["xp"] = xp;
    jsonObj["type"] = type;
    jsonObj["level"] = level;
    jsonObj["baseAttackSpeed"] = baseAttackSpeed;
    jsonObj["baseDamagePerShot"] = baseDamagePerShoot;
    jsonObj["xpForLevelUp"] = xpForLevelUp;
    jsonObj["i"] = i;
    jsonObj["j"] = j;
}

void Tower::read(const QJsonObject &jsonObj)
{
    id = jsonObj["id"].toString();
    damagePerShoot = jsonObj["damagePerShot"].toInt();
    attackSpeed = jsonObj["attackSpeed"].toInt();
    xp = jsonObj["xp"].toInt();
    type = jsonObj["type"].toInt();
    level = jsonObj["level"].toInt();
    baseAttackSpeed = jsonObj["baseAttackSpeed"].toInt();
    baseDamagePerShoot = jsonObj["baseDamagePerShot"].toInt();
    xpForLevelUp = jsonObj["xpForLevelUp"].toInt();
    i = jsonObj["i"].toInt();
    j = jsonObj["j"].toInt();
}

QString Tower::getId() const
{
    return id;
}

void Tower::setId(const QString &value)
{
    id = value;
}

int Tower::getDamagePerShoot() const
{
    return damagePerShoot;
}

void Tower::setDamagePerShoot(int value)
{
    damagePerShoot = value;
}

int Tower::getAttackSpeed() const
{
    return attackSpeed;
}

void Tower::setAttackSpeed(int value)
{
    attackSpeed = value;
}

int Tower::getI() const
{
    return i;
}

void Tower::setI(int value)
{
    i = value;
}

int Tower::getJ() const
{
    return j;
}

void Tower::setJ(int value)
{
    j = value;
}

int Tower::getXp() const{
    return xp;
}

void Tower::setXp(int value){
    xp = value;
}

int Tower::getXpForLevelUp() const{
    return xpForLevelUp;
}

void Tower::setXpForLevelUp(int value){
    xpForLevelUp = value;
}

int Tower::getType() const{
    return type;
}

void Tower::setType(int value){
    type = value;
}

File: src/ce/itcr/gladiators/genetic/resources/Tower.java
package ce.itcr.gladiators.genetic.resources;

public class Tower {
	
	private int attackSpeed;
	private int damagePerShot;
	private int level;
	private int type;
	private int xpForLevelUp;
	private int xp;
	private int baseAttackSpeed;
	private int baseDamagePerShot;
	private int i;
	private int j;

	public int getBaseAttackSpeed() {
		return baseAttackSpeed;
	}

	public void setBaseAttackSpeed(int baseAttackSpeed) {
		this.baseAttackSpeed = baseAttackSpeed;
	}

	public int getBaseDamagePerShot() {
		return baseDamagePerShot;
	}

	public void setBaseDamagePerShot(int baseDamagePerShot) {
		this.baseDamagePerShot = baseDamagePerShot;
	}

	public int getI() {
		return i;
	}

	public void setI(int i) {
		this.i = i;
	}

	public int getJ() {
		return j;
	}

	public void setJ(int j) {
		this.j = j;
	}

	public Tower() {
		super();
		this.attackSpeed = 1;
		this.damagePerShot = 1;
		this.baseAttackSpeed = 1;
		this.baseDamagePerShot = 1;
		this.level = 1;
		this.type = 1;
		this.xpForLevelUp = 10;
		this.xp = 0;
		this.i = -1;
		this.j = -1;
	}

	public Tower(int attackSpeed, int damagePerShot, int level, int type, int xpForLevelUp, int xp, int baseAttackSpeed, int baseDamagePerShot, int i, int j) {
		this.attackSpeed = attackSpeed;
		this.damagePerShot = damagePerShot;
		this.level = level;
		this.type = type;
		this.xpForLevelUp = xpForLevelUp;
		this.xp = xp;
		this.baseAttackSpeed = baseAttackSpeed;
		this.baseDamagePerShot = baseDamagePerShot;
		this.i = i;
		this.j = j;
	}

	public void levelUp() {
		if (this.level == 4 && this.type < 3) {
			this.type++;
			this.level = 1;
			this.xp -= xpForLevelUp;
			this.xpForLevelUp += 10;
			upgradeRNG();
		}
		if (this.attackSpeed > 500);
			this.attackSpeed -= 10;
		this.damagePerShot += 5;
		this.level++;
		this.xp -= this.xpForLevelUp;
		this.xpForLevelUp += 10;
	}
	
	private void upgradeRNG() {
		if (Math.random() < 0.1) {
			this.baseDamagePerShot = this.damagePerShot;
			this.baseAttackSpeed = this.attackSpeed;
			return;
		}
		this.attackSpeed = this.baseAttackSpeed;
		this.damagePerShot = this.baseDamagePerShot;
	}

	public int getAttackSpeed() {
		return attackSpeed;
	}

	public void setAttackSpeed(int attackSpeed) {
		this.attackSpeed = attackSpeed;
	}

	public int getDamagePerShot() {
		return damagePerShot;
	}

	public void setDamagePerShot(int damagePerShot) {
		this.damagePerShot = damagePerShot;
	}

	public int getLevel() {
		return level;
	}

	public void setLevel(int level) {
		this.level = level;
	}

	public int getType() {
		return type;
	}

	public void setType(int type) {
		this.type = type;
	}

	public int getXpForLevelUp() {
		return xpForLevelUp;
	}

	public void setXpForLevelUp(int xpForLevelUp) {
		this.xpForLevelUp = xpForLevelUp;
	}

	public int getXp() {
		return xp;
	}

	public void setXp(int xp) {
		this.xp = xp;
	}
	
	
	
}